pkg/proxy/proxy.go

```diff
@@ -264,51 +266,221 @@ func (p *Proxy) health(ctx echo.Context) error {
-		workspace, err := handlers.GetUserWorkspace(ctx, p.spaceLister, workspaceName)
+	// list all workspaces the user has access to
+	workspaces, err := handlers.ListUserWorkspaces(ctx, p.spaceLister)
+	if err != nil {
+		return nil, crterrors.NewInternalError(errs.New("unable to retrieve user workspaces"), err.Error())
+	}
+
+	// check whether the user's has access to the home workspace
+	// and whether the requestedNamespace -if any- exists in the workspace.
+	requestedNamespace := namespaceFromCtx(ctx)
+	if err := validateWorkspaceRequest("", requestedNamespace, workspaces); err != nil {
+		return nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
+	}
+
+	// return the cluster access
+	return cluster, nil
+}
+
+// processWorkspaceRequest process an HTTP Request targeting a specific workspace.
+func (p *Proxy) processWorkspaceRequest(ctx echo.Context, userID, username, workspaceName, proxyPluginName string) (*access.ClusterAccess, error) {
+	// check if the user has access to the target workspace
+	if err := p.validateSpaceAccess(ctx, userID, username, workspaceName); err != nil {
+		return nil, err
+	}
+
+	// before proxying the request, verify that the user has a spacebinding for
+	// the workspace and that the namespace -if any- belongs to the workspace
+	workspaces, err := p.listUserWorkspaces(ctx, workspaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// check whether the user's has access to the home workspace
+	// and whether the requestedNamespace -if any- exists in the workspace.
+	requestedNamespace := namespaceFromCtx(ctx)
+	if err := validateWorkspaceRequest(workspaceName, requestedNamespace, workspaces); err != nil {
+		return nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
+	}
+
+	// retrieve the ClusterAccess for the user and the target workspace
+	return p.getClusterAccess(ctx, userID, username, workspaceName, proxyPluginName, workspaces)
+}
+
+// validateSpaceAccess checks whether the user has access to the target workspace and that the Space exists.
+func (p *Proxy) validateSpaceAccess(ctx echo.Context, userID, username, workspaceName string) error {
+	if err := p.validateUserAccess(ctx, userID, username); err != nil {
+		return crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
+	}
+	if err := p.validateSpace(workspaceName); err != nil {
+		return crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
+	}
+
+	return nil
+}
+
+// validateSpace checks whether the Space exists.
+func (p *Proxy) validateSpace(workspaceName string) error {
+	if _, err := p.app.InformerService().GetSpace(workspaceName); err != nil {
+		// log the actual error but do not return it so that it doesn't reveal information about a space that may not belong to the requestor
+		log.Error(nil, err, "unable to get target cluster for workspace "+workspaceName)
```

Comment:
shouldn't this be something like "the requested space is not available", as the error we return ? the check doesn't seem related to the target cluster.

Reply:
fixed in 8d944fb19e69a25eb1e4a168fc56ae3cdcf10969